lib/active_merchant/billing/gateways/paymentez.rb

```diff
@@ -0,0 +1,274 @@
+
+      self.supported_countries = %w[MX EC VE CO BR CL]
+      self.default_currency = 'USD'
+      self.supported_cardtypes = %i[visa master american_express diners_club]
+
+      self.homepage_url = 'https://secure.paymentez.com/'
+      self.display_name = 'Paymentez'
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        1 => :processing_error,
+        6 => :card_declined,
+        9 => :card_declined,
+        10 => :processing_error,
+        11 => :card_declined,
+        12 => :config_error,
+        13 => :config_error,
+        19 => :invalid_cvc,
+        20 => :config_error,
+        21 => :card_declined,
+        22 => :card_declined,
+        23 => :card_declined,
+        24 => :card_declined,
+        25 => :card_declined,
+        26 => :card_declined,
+        27 => :card_declined,
+        28 => :card_declined
+      }.freeze
+
+      CARD_MAPPING = {
+        'visa' => 'vi',
+        'master' => 'mc',
+        'american_express' => 'ax',
+        'diners_club' => 'di'
+      }.freeze
+
+      def initialize(options = {})
+        requires!(options, :application_code, :app_key)
+        super
+      end
+
+      def purchase(money, payment, options = {})
+        post = {}
+
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_customer_data(post, options)
+
+        commit_transaction('debit_cc', post)
+      end
+
+      def authorize(money, payment, options = {})
+        post = {}
+
+        add_invoice(post, money, options)
+        add_customer_data(post, options)
+
+        MultiResponse.run do |r|
+          r.process { store(payment, options) }
+          post[:card] = { token: r.authorization }
+          r.process { commit_transaction('authorize', post) }
+        end
+      end
+
+      def capture(_money, authorization, _options = {})
+        post = { transaction: { id: authorization } }
+
+        commit_transaction('capture', post)
+      end
+
+      def refund(_money, authorization, options = {})
+        void(authorization, options)
+      end
+
+      def void(authorization, _options = {})
+        post = { transaction: { id: authorization } }
+        commit_transaction('refund', post)
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process { void(r.authorization, options) }
+        end
+      end
+
+      def store(credit_card, options = {})
+        post = {}
+
+        add_customer_data(post, options)
+        add_payment(post, credit_card)
+
+        response = commit_card('add', post)
+        if !response.success? && !(token = extract_previous_card_token(response)).nil?
+          unstore("#{token}|#{post[:user][:id]}")
+          response = commit_card('add', post)
+        end
+        response
+      end
+
+      def unstore(identification, _options = {})
+        card_token, user_id = identification.split('|', 2)
```

Comment:
I think we might prefer to pull the user_id from options instead of requiring the user to concatenate it to the token when passing it in here.

Reply:
Oh, point. Sorry, first interaction with TPV. That'll also avoid me having to make `store` munge. Thanks!